hivemind/dht/node.py

```diff
@@ -368,11 +376,13 @@ def _update_cache_on_store(self, key_id: DHTID, subkeys: List[Subkey], binary_va
             self.protocol.cache.store(key_id, stored_value_bytes, stored_expiration)
         elif not store_succeeded and not is_dictionary:  # store rejected, check if local cache is also obsolete
             rejected_value, rejected_expiration = max(zip(binary_values, expirations), key=lambda p: p[1])
-            if (self.protocol.cache.get(key_id)[1] or float("inf")) <= rejected_expiration:  # cache would be rejected
+            if (self.protocol.cache.get(key_id) or (None, float("inf")))[1] <= rejected_expiration:  # cache would be rejected
                 self._schedule_for_refresh(key_id, refresh_time=get_dht_time())  # fetch new key in background (asap)
         elif is_dictionary and key_id in self.protocol.cache:  # there can be other keys and we should update
-            for subkey, stored_value_bytes, expiration_time in zip(subkeys, binary_values, expirations):
-                self.protocol.cache.store_subkey(key_id, subkey, stored_value_bytes, expiration_time)
+            for subkey, stored_value_bytes, expiration_time, valid in zip(
+                    subkeys, binary_values, expirations, store_ok):
+                if valid:
+                    self.protocol.cache.store_subkey(key_id, subkey, stored_value_bytes, expiration_time)
```

Comment:
```suggestion
            for subkey, stored_value_bytes, expiration_time, accepted in zip(
                    subkeys, binary_values, expirations, store_ok):
                if accepted:
                    self.protocol.cache.store_subkey(key_id, subkey, stored_value_bytes, expiration_time)
```

[Optional] i would argue that valid=False does not necessarily mean that the record is invalid: peers may have refused it for a number of reasons, such as having a record with newer expiration time.